### ISSUE TITLE
What did you think about the process of writing a command?

### ISSUE THREAD

**Issue by danielbachhuber** [trendwerk/faker#2]
What went well, and what could be improved? Any documentation changes you'd suggest?

**Comment by haroldangenent** [comment]
The overall experience on writing the command was pretty good. The hardest part was probably figuring out how `wp-cli` knows about what commands exist.

**Approach**
It took me some time to get things up & running. I slowly figured out I had these approaches to building a command:
- Create a WordPress plugin that checks for `WP_CLI`
- Standalone package for `wp-cli`

I preferred the second option, but it seemed a bit hidden in the docs at first. I only found this information because I wanted to create a simple Composer package (that could easily request it's own dependencies) instead of a WordPress plugin.

Of course, [the Distribution chapter](https://wp-cli.org/docs/commands-cookbook/#distribution) describes how the command could be distributed, but I found that this is one of the early choices on how to start developing the command.

To get started, I mostly looked around in other existing packages that seemed to use PSR-4 autoloading and dependencies to get a better grasp for the internals of `wp package`.

**`wp scaffold package`**
Ran the scaffolder, [deleted the tests](https://github.com/trendwerk/faker/issues/3) (sorry for that, but wanted to start as simple as possible and tests are one of my shortcomings I'm planning to work on) and ran: `wp hello-world`.

This didn't work, the command was not available. I didn't really expect this (and agree with https://github.com/wp-cli/scaffold-package-command/issues/32). After that I dug around in how the packages worked internally and figured they also work through Composer (yay).

Apparently, my scaffolded plugin was located in `vendor/`. I moved it to `src/<Namespace>` and added the file and PSR-4 to Composer's autoloader option (in the Packages folder). That way the dependencies would also be included, without changing the package code to manually include `vendor/autoload.php` if it existed. Ran `composer dump-autoload` and `wp hello-world` could be used. I'm not sure if this is the preferred method, but couldn't find much about this in the docs. 

I still had some questions left unanswered, like "(how) are my dependencies fetched by `wp package`?" and "will the autoloaders (file and PSR-4) work properly?", but figured this would work like any Composer package and they did.

**Basics: done**
When the basic command was done, developing for `wp-cli` was simple and straightforward. The [Internal API](https://wp-cli.org/docs/internal-api/) was clear and easy to understand.

Relying heavily on external dependencies, it was really nice that the packages work through a Composer repository.

**Versioning**
Sidenote: I'm not sure what versioning approach is recommended. I'm used to semver, but a lot of other packages are using `dev-master` and the minimum stability seems to be `dev`. What would you recommend?

**TLDR; My main concern was having to dig in to the internals to start development. I think this could be a threshold for a lot of WordPress developers. Development itself was great. I'd suggest adding a chapter about creating a `wp-cli` package, that describes the workflow.**

_By the way, great effort for asking feedback from developers. I hope you understand a bit more about how this process went for me and find it useful._

**Comment by danielbachhuber** [comment]
This is a great writeup, @haroldangenent. I appreciate you taking the time to go into such detail.

I've created an issue to improve some of the Commands Cookbook documentation: https://github.com/wp-cli/wp-cli/issues/3323

> This didn't work, the command was not available. I didn't really expect this

Sorry about this. I agree this is a huge pain point in creating a new command right now. At the very least, it should be documented. If you have a relatively strong knowledge of Composer though, I'd appreciate your help in working towards a more robust solution: https://github.com/wp-cli/wp-cli/issues/2565

> Sidenote: I'm not sure what versioning approach is recommended. I'm used to semver, but a lot of other packages are using dev-master and the minimum stability seems to be dev. What would you recommend?

Generally speaking, semver. Given we're in the early days of packages though, the concept of versioning and updates is somewhat secondary (e.g. https://github.com/wp-cli/wp-cli/issues/2544). I'm also trying to get buy-in on forming community standards for WP-CLI package best practices: https://github.com/wp-cli/wp-cli/issues/3197

**Comment by haroldangenent** [comment]
@danielbachhuber Thanks for the suggestion. Just released `0.1.0`. However, it seems like `wp-cli` is only requiring `dev-master` upon installation. Is that correct?

By the way, I noticed #5 but I really cannot find the source of this issue (dug around a lot). It seems that packages like `aaemnnosttv/wp-cli-dotenv-command` also have this error.

**Comment by haroldangenent** [comment]
@danielbachhuber Also noticed that `wp package uninstall` removes the package, but not it's dependencies. Should I create an issue about this or is this meant behaviour?

_(`composer remove vendor/package` does remove it's dependencies)._

**Comment by danielbachhuber** [comment]
> However, it seems like wp-cli is only requiring dev-master upon installation. Is that correct?

You can optionally provide a version constraint with `--version=<version>`. However, the default is to use `dev-master`.

> Also noticed that `wp package uninstall` removes the package, but not it's dependencies. Should I create an issue about this or is this meant behaviour?

This seems like a bug. Creating an issue for it would be great — thanks!
